<a href="https://colab.research.google.com/github/tinkercademy/ml-notebooks/blob/main/04_Pandas_and_Dataframes_Pokemon_(Answers).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Pandas and Dataframes

**Pandas** is a popular Python library, used for statistical analysis. We operate on **dataframes**, which you can think of as "Excel sheets as variables in Python".

In [1]:
import pandas as pd

In [2]:
# If you are mounting from google drive, run this. Else, you can ignore this
from google.colab import drive
drive.mount('/content/drive')

MessageError: ignored

## Importing data

Now, let's import some data! We do this using the `read_csv` function within the `pd` library, so we call it with `pd.read_csv`. We tell it one thing: the location of the file to be read.

Let's get started by importing a CSV file, which are plain-text versions of data values! Some sample CSV files can be found in this server's `Resources` folder. To use them, you'll need to go to the homepage, download the CSV files from the `Resources` folder, and **re-upload to the same folder as this notebook**.

Once you're ready, run this line to store all that data in to a single variable:

In [3]:
data = pd.read_csv('https://raw.githubusercontent.com/iamamangosteen/pythonforbusiness/master/pokedex_(Update_2020.05).csv', index_col = "name")

In [4]:
data

pokedex_number german_name  \
name                                                         
Bulbasaur                                    1     Bisasam   
Ivysaur                                      2   Bisaknosp   
Venusaur                                     3    Bisaflor   
Mega Venusaur                                3    Bisaflor   
Charmander                                   4    Glumanda   
...                                        ...         ...   
Zacian Hero of Many Battles                888         NaN   
Zamazenta Crowned Shield                   889         NaN   
Zamazenta Hero of Many Battles             889         NaN   
Eternatus                                  890         NaN   
Eternatus Eternamax                        890         NaN   

                                      japanese_name  generation     status  \
name                                                                         
Bulbasaur                       フシギダネ (Fushigidane)           1     Normal   
Ivysaur                          フシギソウ (Fushigisou)           1     Normal   
Venusaur                        フシギバナ (Fushigibana)           1     Normal   
Mega Venusaur                   フシギバナ (Fushigibana)           1     Normal   
Charmander                          ヒトカゲ (Hitokage)           1     Normal   
...                                             ...         ...        ...   
Zacian Hero of Many Battles                     NaN           8  Legendary   
Zamazenta Crowned Shield                        NaN           8  Legendary   
Zamazenta Hero of Many Battles                  NaN           8  Legendary   
Eternatus                                       NaN           8  Legendary   
Eternatus Eternamax                             NaN           8  Legendary   

                                         species  type_number    type_1  \
name                                                                      
Bulbasaur                           Seed Pokémon            2     Grass   
Ivysaur                             Seed Pokémon            2     Grass   
Venusaur                            Seed Pokémon            2     Grass   
Mega Venusaur                       Seed Pokémon            2     Grass   
Charmander                        Lizard Pokémon            1      Fire   
...                                          ...          ...       ...   
Zacian Hero of Many Battles      Warrior Pokémon            1     Fairy   
Zamazenta Crowned Shield         Warrior Pokémon            2  Fighting   
Zamazenta Hero of Many Battles   Warrior Pokémon            1  Fighting   
Eternatus                       Gigantic Pokémon            2    Poison   
Eternatus Eternamax             Gigantic Pokémon            2    Poison   

                                type_2  height_m  ...  against_ground  \
name                                              ...                   
Bulbasaur                       Poison       0.7  ...             1.0   
Ivysaur                         Poison       1.0  ...             1.0   
Venusaur                        Poison       2.0  ...             1.0   
Mega Venusaur                   Poison       2.4  ...             1.0   
Charmander                         NaN       0.6  ...             2.0   
...                                ...       ...  ...             ...   
Zacian Hero of Many Battles        NaN       2.8  ...             1.0   
Zamazenta Crowned Shield         Steel       2.9  ...             2.0   
Zamazenta Hero of Many Battles     NaN       2.9  ...             1.0   
Eternatus                       Dragon      20.0  ...             2.0   
Eternatus Eternamax             Dragon     100.0  ...             2.0   

                                against_flying against_psychic against_bug  \
name                                                                         
Bulbasaur                                  2.0             2.0        1.00   
Ivysaur                                    2.0    

Printing is a bit unwieldy because of all the data. To read just a bit of info from the beginning, which lets you check if everything imported correctly, use `.head()`:

In [5]:
data.head()

pokedex_number german_name        japanese_name  generation  \
name                                                                         
Bulbasaur                   1     Bisasam  フシギダネ (Fushigidane)           1   
Ivysaur                     2   Bisaknosp   フシギソウ (Fushigisou)           1   
Venusaur                    3    Bisaflor  フシギバナ (Fushigibana)           1   
Mega Venusaur               3    Bisaflor  フシギバナ (Fushigibana)           1   
Charmander                  4    Glumanda      ヒトカゲ (Hitokage)           1   

               status         species  type_number type_1  type_2  height_m  \
name                                                                          
Bulbasaur      Normal    Seed Pokémon            2  Grass  Poison       0.7   
Ivysaur        Normal    Seed Pokémon            2  Grass  Poison       1.0   
Venusaur       Normal    Seed Pokémon            2  Grass  Poison       2.0   
Mega Venusaur  Normal    Seed Pokémon            2  Grass  Poison       2.4   
Charmander     Normal  Lizard Pokémon            1   Fire     NaN       0.6   

               ...  against_ground  against_flying against_psychic  \
name           ...                                                   
Bulbasaur      ...             1.0             2.0             2.0   
Ivysaur        ...             1.0             2.0             2.0   
Venusaur       ...             1.0             2.0             2.0   
Mega Venusaur  ...             1.0             2.0             2.0   
Charmander     ...             2.0             1.0             1.0   

              against_bug against_rock  against_ghost  against_dragon  \
name                                                                    
Bulbasaur             1.0          1.0            1.0             1.0   
Ivysaur               1.0          1.0            1.0             1.0   
Venusaur              1.0          1.0            1.0             1.0   
Mega Venusaur         1.0          1.0            1.0             1.0   
Charmander            0.5          2.0            1.0             1.0   

               against_dark  against_steel  against_fairy  
name                                                       
Bulbasaur               1.0            1.0            0.5  
Ivysaur                 1.0            1.0            0.5  
Venusaur                1.0            1.0            0.5  
Mega Venusaur           1.0            1.0            0.5  
Charmander              1.0            0.5            0.5  

[5 rows x 49 columns]

Notice that the column headings are each of the columns in the CSV file, and the row headings are just numbers. This is OK, but not quite right--we want each row heading to be the pokemon name. In pandas terminology, we want to **set the pokemon name column as the index**. To do this, we re-import, and specify an `index_col`:

In [6]:
data = pd.read_csv('https://raw.githubusercontent.com/iamamangosteen/pythonforbusiness/master/pokedex_(Update_2020.05).csv', index_col = 'name')

In [7]:
data.head()

pokedex_number german_name        japanese_name  generation  \
name                                                                         
Bulbasaur                   1     Bisasam  フシギダネ (Fushigidane)           1   
Ivysaur                     2   Bisaknosp   フシギソウ (Fushigisou)           1   
Venusaur                    3    Bisaflor  フシギバナ (Fushigibana)           1   
Mega Venusaur               3    Bisaflor  フシギバナ (Fushigibana)           1   
Charmander                  4    Glumanda      ヒトカゲ (Hitokage)           1   

               status         species  type_number type_1  type_2  height_m  \
name                                                                          
Bulbasaur      Normal    Seed Pokémon            2  Grass  Poison       0.7   
Ivysaur        Normal    Seed Pokémon            2  Grass  Poison       1.0   
Venusaur       Normal    Seed Pokémon            2  Grass  Poison       2.0   
Mega Venusaur  Normal    Seed Pokémon            2  Grass  Poison       2.4   
Charmander     Normal  Lizard Pokémon            1   Fire     NaN       0.6   

               ...  against_ground  against_flying against_psychic  \
name           ...                                                   
Bulbasaur      ...             1.0             2.0             2.0   
Ivysaur        ...             1.0             2.0             2.0   
Venusaur       ...             1.0             2.0             2.0   
Mega Venusaur  ...             1.0             2.0             2.0   
Charmander     ...             2.0             1.0             1.0   

              against_bug against_rock  against_ghost  against_dragon  \
name                                                                    
Bulbasaur             1.0          1.0            1.0             1.0   
Ivysaur               1.0          1.0            1.0             1.0   
Venusaur              1.0          1.0            1.0             1.0   
Mega Venusaur         1.0          1.0            1.0             1.0   
Charmander            0.5          2.0            1.0             1.0   

               against_dark  against_steel  against_fairy  
name                                                       
Bulbasaur               1.0            1.0            0.5  
Ivysaur                 1.0            1.0            0.5  
Venusaur                1.0            1.0            0.5  
Mega Venusaur           1.0            1.0            0.5  
Charmander              1.0            0.5            0.5  

[5 rows x 49 columns]

Now, the first two rows look a bit weird, but that's pandas' way of telling us that "name" is an index column. You can have multiple index columns, but that's beyond the scope of this class.

## Some DataFrame operations

Some useful functions you can do with DataFrames below. Try them out, and see what they do!

In [8]:
data.max()

<ipython-input-8-8637789457d2>:1: FutureWarning: The default value of numeric_only in DataFrame.max is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.max()


pokedex_number                    890
generation                          8
status                  Sub Legendary
species             Zen Charm Pokémon
type_number                         2
type_1                          Water
height_m                        100.0
weight_kg                       999.9
abilities_number                    3
total_points                     1125
hp                                255
attack                            190
defense                           250
sp_attack                         194
sp_defense                        250
speed                             180
catch_rate                      255.0
base_friendship                 140.0
base_experience                 608.0
egg_type_number                     2
percentage_male                 100.0
egg_cycles                      120.0
against_normal                    1.0
against_fire                      4.0
against_water                     4.0
against_electric                  4.0
against_gras

In [9]:
data.min()

<ipython-input-9-9129c1932162>:1: FutureWarning: The default value of numeric_only in DataFrame.min is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.min()


pokedex_number                      1
generation                          1
status                      Legendary
species             Abundance Pokémon
type_number                         1
type_1                            Bug
height_m                          0.1
weight_kg                         0.1
abilities_number                    0
total_points                      175
hp                                  1
attack                              5
defense                             5
sp_attack                          10
sp_defense                         20
speed                               5
catch_rate                        3.0
base_friendship                   0.0
base_experience                  36.0
egg_type_number                     0
percentage_male                   0.0
egg_cycles                        5.0
against_normal                    0.0
against_fire                      0.0
against_water                     0.0
against_electric                  0.0
against_gras

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1028 entries, Bulbasaur to Eternatus Eternamax
Data columns (total 49 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   pokedex_number    1028 non-null   int64  
 1   german_name       938 non-null    object 
 2   japanese_name     938 non-null    object 
 3   generation        1028 non-null   int64  
 4   status            1028 non-null   object 
 5   species           1028 non-null   object 
 6   type_number       1028 non-null   int64  
 7   type_1            1028 non-null   object 
 8   type_2            542 non-null    object 
 9   height_m          1028 non-null   float64
 10  weight_kg         1027 non-null   float64
 11  abilities_number  1028 non-null   int64  
 12  ability_1         1025 non-null   object 
 13  ability_2         513 non-null    object 
 14  ability_hidden    810 non-null    object 
 15  total_points      1028 non-null   int64  
 16  hp                1028 n

In [11]:
data.describe()

pokedex_number   generation  type_number     height_m    weight_kg  \
count     1028.000000  1028.000000  1028.000000  1028.000000  1027.000000   
mean       437.711089     4.034047     1.527237     1.368093    69.753749   
std        259.366480     2.234937     0.499501     3.380126   129.221230   
min          1.000000     1.000000     1.000000     0.100000     0.100000   
25%        213.750000     2.000000     1.000000     0.600000     8.800000   
50%        433.500000     4.000000     2.000000     1.000000    28.500000   
75%        663.250000     6.000000     2.000000     1.500000    69.100000   
max        890.000000     8.000000     2.000000   100.000000   999.900000   

       abilities_number  total_points           hp       attack      defense  \
count       1028.000000   1028.000000  1028.000000  1028.000000  1028.000000   
mean           2.284047    437.571984    69.577821    80.119650    74.475681   
std            0.794981    121.664910    26.385849    32.372321    31.303309   
min            0.000000    175.000000     1.000000     5.000000     5.000000   
25%            2.000000    330.000000    50.000000    55.000000    50.000000   
50%            2.000000    455.000000    66.500000    76.000000    70.000000   
75%            3.000000    510.000000    80.000000   100.000000    90.000000   
max            3.000000   1125.000000   255.000000   190.000000   250.000000   

       ...  against_ground  against_flying  against_psychic  against_bug  \
count  ...     1028.000000     1028.000000      1028.000000  1028.000000   
mean   ...        1.084630        1.166342         0.979329     0.992461   
std    ...        0.784937        0.593030         0.499146     0.598301   
min    ...        0.000000        0.250000         0.000000     0.000000   
25%    ...        0.500000        1.000000         1.000000     0.500000   
50%    ...        1.000000        1.000000         1.000000     1.000000   
75%    ...        1.625000        1.000000         1.000000     1.000000   
max    ...        4.000000        4.000000         4.000000     4.000000   

       against_rock  against_ghost  against_dragon  against_dark  \
count   1028.000000    1028.000000     1028.000000   1028.000000   
mean       1.239786       1.010700        0.975681      1.065661   
std        0.699156       0.558533        0.377549      0.451054   
min        0.250000       0.000000        0.000000      0.250000   
25%        1.000000       1.000000        1.000000      1.000000   
50%        1.000000       1.000000        1.000000      1.000000   
75%        2.000000       1.000000        1.000000      1.000000   
max        4.000000       4.000000        2.000000      4.000000   

       against_steel  against_fairy  
count    1028.000000    1028.000000  
mean        0.980302       1.084874  
std         0.503434       0.527743  
min         0.000000       0.000000  
25%         0.500000       1.000000  
50%         1.000000       1.000000  
75%         1.000000       1.000000  
max         4.000000       4.000000  

[8 rows x 37 columns]

In [12]:
data.columns

Index(['pokedex_number', 'german_name', 'japanese_name', 'generation',
       'status', 'species', 'type_number', 'type_1', 'type_2', 'height_m',
       'weight_kg', 'abilities_number', 'ability_1', 'ability_2',
       'ability_hidden', 'total_points', 'hp', 'attack', 'defense',
       'sp_attack', 'sp_defense', 'speed', 'catch_rate', 'base_friendship',
       'base_experience', 'growth_rate', 'egg_type_number', 'egg_type_1',
       'egg_type_2', 'percentage_male', 'egg_cycles', 'against_normal',
       'against_fire', 'against_water', 'against_electric', 'against_grass',
       'against_ice', 'against_fight', 'against_poison', 'against_ground',
       'against_flying', 'against_psychic', 'against_bug', 'against_rock',
       'against_ghost', 'against_dragon', 'against_dark', 'against_steel',
       'against_fairy'],
      dtype='object')

In [13]:
data.index

Index(['Bulbasaur', 'Ivysaur', 'Venusaur', 'Mega Venusaur', 'Charmander',
       'Charmeleon', 'Charizard', 'Mega Charizard X', 'Mega Charizard Y',
       'Squirtle',
       ...
       'Duraludon', 'Dreepy', 'Drakloak', 'Dragapult', 'Zacian Crowned Sword',
       'Zacian Hero of Many Battles', 'Zamazenta Crowned Shield',
       'Zamazenta Hero of Many Battles', 'Eternatus', 'Eternatus Eternamax'],
      dtype='object', name='name', length=1028)

In [14]:
data.T

name                        Bulbasaur             Ivysaur  \
pokedex_number                      1                   2   
german_name                   Bisasam           Bisaknosp   
japanese_name     フシギダネ (Fushigidane)  フシギソウ (Fushigisou)   
generation                          1                   1   
status                         Normal              Normal   
species                  Seed Pokémon        Seed Pokémon   
type_number                         2                   2   
type_1                          Grass               Grass   
type_2                         Poison              Poison   
height_m                          0.7                 1.0   
weight_kg                         6.9                13.0   
abilities_number                    2                   2   
ability_1                    Overgrow            Overgrow   
ability_2                         NaN                 NaN   
ability_hidden            Chlorophyll         Chlorophyll   
total_points                      318                 405   
hp                                 45                  60   
attack                             49                  62   
defense                            49                  63   
sp_attack                          65                  80   
sp_defense                         65                  80   
speed                              45                  60   
catch_rate                       45.0                45.0   
base_friendship                  70.0                70.0   
base_experience                  64.0               142.0   
growth_rate               Medium Slow         Medium Slow   
egg_type_number                     2                   2   
egg_type_1                      Grass               Grass   
egg_type_2                    Monster             Monster   
percentage_male                  87.5                87.5   
egg_cycles                       20.0                20.0   
against_normal                    1.0                 1.0   
against_fire                      2.0                 2.0   
against_water                     0.5                 0.5   
against_electric                  0.5                 0.5   
against_grass                    0.25                0.25   
against_ice                       2.0                 2.0   
against_fight                     0.5                 0.5   
against_poison                    1.0                 1.0   
against_ground                    1.0                 1.0   
against_flying                    2.0                 2.0   
against_psychic                   2.0                 2.0   
against_bug                       1.0                 1.0   
against_rock                      1.0                 1.0   
against_ghost                     1.0                 1.0   
against_dragon                    1.0                 1.0   
against_dark                      1.0                 1.0   
against_steel                     1.0                 1.0   
against_fairy                     0.5                 0.5   

name                         Venusaur        Mega Venusaur       Charmander  \
pokedex_number                      3                    3                4   
german_name                  Bisaflor             Bisaflor         Glumanda   
japanese_name     フシギバナ (Fushigibana)  フシギバナ (Fushigibana)  ヒトカゲ (Hitokage)   
generation                          1                    1                1   
status                         Normal               Normal           Normal   
species                  Seed Pokémon         Seed Pokémon   Lizard Pokémon   
type_number                         2                    2                1   
type_1                          Grass                Grass             Fire   
type_2                         Poison               Poison              NaN   
height_m                          2.0                  2.4              0.6   
weight_kg                       100.0                155.5              8.5   
a

### <font color="red">Exercise 1: Import and check

New pokedex data was just released! Import the data from the new CSV (from Apr 2021) and check how many rows of data it has.

In [15]:
data = pd.read_csv('https://raw.githubusercontent.com/iamamangosteen/pythonforbusiness/master/pokedex_(Update_2021.04).csv', index_col = 'name')
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1045 entries, Bulbasaur to Calyrex Shadow Rider
Data columns (total 49 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   pokedex_number    1045 non-null   int64  
 1   german_name       1045 non-null   object 
 2   japanese_name     1045 non-null   object 
 3   generation        1045 non-null   int64  
 4   status            1045 non-null   object 
 5   species           1045 non-null   object 
 6   type_number       1045 non-null   int64  
 7   type_1            1045 non-null   object 
 8   type_2            553 non-null    object 
 9   height_m          1045 non-null   float64
 10  weight_kg         1044 non-null   float64
 11  abilities_number  1045 non-null   int64  
 12  ability_1         1042 non-null   object 
 13  ability_2         516 non-null    object 
 14  ability_hidden    813 non-null    object 
 15  total_points      1045 non-null   int64  
 16  hp                1045 

<hr>

## Reading and filtering data in DataFrames

There are many, many ways of accessing data in DataFrames. Here are a few ways--you can read up on other ways at the [Pandas DataFrame documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) page. It's worth clicking over just to see the variety of functions you can call to handle DataFrames!

Here, though, we'll start with accessing information the way you'd expect, by row and column:

In [16]:
data.head()

pokedex_number german_name        japanese_name  generation  \
name                                                                         
Bulbasaur                   1     Bisasam  フシギダネ (Fushigidane)           1   
Ivysaur                     2   Bisaknosp   フシギソウ (Fushigisou)           1   
Venusaur                    3    Bisaflor  フシギバナ (Fushigibana)           1   
Mega Venusaur               3    Bisaflor  フシギバナ (Fushigibana)           1   
Charmander                  4    Glumanda      ヒトカゲ (Hitokage)           1   

               status         species  type_number type_1  type_2  height_m  \
name                                                                          
Bulbasaur      Normal    Seed Pokémon            2  Grass  Poison       0.7   
Ivysaur        Normal    Seed Pokémon            2  Grass  Poison       1.0   
Venusaur       Normal    Seed Pokémon            2  Grass  Poison       2.0   
Mega Venusaur  Normal    Seed Pokémon            2  Grass  Poison       2.4   
Charmander     Normal  Lizard Pokémon            1   Fire     NaN       0.6   

               ...  against_ground  against_flying against_psychic  \
name           ...                                                   
Bulbasaur      ...             1.0             2.0             2.0   
Ivysaur        ...             1.0             2.0             2.0   
Venusaur       ...             1.0             2.0             2.0   
Mega Venusaur  ...             1.0             2.0             2.0   
Charmander     ...             2.0             1.0             1.0   

              against_bug against_rock  against_ghost  against_dragon  \
name                                                                    
Bulbasaur             1.0          1.0            1.0             1.0   
Ivysaur               1.0          1.0            1.0             1.0   
Venusaur              1.0          1.0            1.0             1.0   
Mega Venusaur         1.0          1.0            1.0             1.0   
Charmander            0.5          2.0            1.0             1.0   

               against_dark  against_steel  against_fairy  
name                                                       
Bulbasaur               1.0            1.0            0.5  
Ivysaur                 1.0            1.0            0.5  
Venusaur                1.0            1.0            0.5  
Mega Venusaur           1.0            1.0            0.5  
Charmander              1.0            0.5            0.5  

[5 rows x 49 columns]

In [17]:
# Row, Column
data.iloc[4,5]

'Lizard Pokémon'

In [18]:
subset = data.iloc[0:200, 1:10]
subset

german_name        japanese_name  generation     status  \
name                                                                    
Bulbasaur         Bisasam  フシギダネ (Fushigidane)           1     Normal   
Ivysaur         Bisaknosp   フシギソウ (Fushigisou)           1     Normal   
Venusaur         Bisaflor  フシギバナ (Fushigibana)           1     Normal   
Mega Venusaur    Bisaflor  フシギバナ (Fushigibana)           1     Normal   
Charmander       Glumanda      ヒトカゲ (Hitokage)           1     Normal   
...                   ...                  ...         ...        ...   
Mega Mewtwo Y       Mewtu       ミュウツー (Mewtwo)           1  Legendary   
Mew                   Mew            ミュウ (Mew)           1   Mythical   
Chikorita         Endivie    チコリータ (Chicorita)           2     Normal   
Bayleef          Lorblatt      ベイリーフ (Bayleaf)           2     Normal   
Meganium          Meganie     メガニウム (Meganium)           2     Normal   

                           species  type_number   type_1  type_2  height_m  
name                                                                        
Bulbasaur             Seed Pokémon            2    Grass  Poison       0.7  
Ivysaur               Seed Pokémon            2    Grass  Poison       1.0  
Venusaur              Seed Pokémon            2    Grass  Poison       2.0  
Mega Venusaur         Seed Pokémon            2    Grass  Poison       2.4  
Charmander          Lizard Pokémon            1     Fire     NaN       0.6  
...                            ...          ...      ...     ...       ...  
Mega Mewtwo Y      Genetic Pokémon            1  Psychic     NaN       1.5  
Mew            New Species Pokémon            1  Psychic     NaN       0.4  
Chikorita             Leaf Pokémon            1    Grass     NaN       0.9  
Bayleef               Leaf Pokémon            1    Grass     NaN       1.2  
Meganium              Herb Pokémon            1    Grass     NaN       1.8  

[200 rows x 9 columns]

In [19]:
data.loc["Bulbasaur":'Charizard', "generation":"type_2"]

generation  status         species  type_number type_1  type_2
name                                                                         
Bulbasaur               1  Normal    Seed Pokémon            2  Grass  Poison
Ivysaur                 1  Normal    Seed Pokémon            2  Grass  Poison
Venusaur                1  Normal    Seed Pokémon            2  Grass  Poison
Mega Venusaur           1  Normal    Seed Pokémon            2  Grass  Poison
Charmander              1  Normal  Lizard Pokémon            1   Fire     NaN
Charmeleon              1  Normal   Flame Pokémon            1   Fire     NaN
Charizard               1  Normal   Flame Pokémon            2   Fire  Flying

In [20]:
rows_to_show = ["Bulbasaur", "Charmander", "Squirtle", "Pikachu"]
cols_to_show = ["type_1", "type_2",'attack', 'defense', 'speed']
data.loc[rows_to_show, cols_to_show]

type_1  type_2  attack  defense  speed
name                                                
Bulbasaur      Grass  Poison      49       49     45
Charmander      Fire     NaN      52       43     65
Squirtle       Water     NaN      48       65     43
Pikachu     Electric     NaN      55       40     90

### <font color="red">Exercise 2: Get data</font>

Import the necessary CSV file, and set up a DataFrame for the base stats 'hp', 'attack', 'speed' for each of the Pokemon 'Bulbasaur', 'Charmander' and 'Squirtle'. Your result should look like this:

| name | hp | attack | speed |
|---|:---:|:---:|:---:|
| Bulbasaur | 45 | 49 | 45 |
| Charmander | 39 | 52 | 65 |
| Squirtle | 44 | 48 | 43 |

In [21]:
rows_to_show = ["Bulbasaur", "Charmander", "Squirtle"]
cols_to_show = ['hp', 'attack', 'speed']

data.loc[rows_to_show, cols_to_show]

hp  attack  speed
name                         
Bulbasaur   45      49     45
Charmander  39      52     65
Squirtle    44      48     43

<hr>

## Filtering data

Here's one of the most powerful features of DataFrames--being able to quickly work with large chunks of data. If you had to do this with for loops, it'd be a bit of pain to filter everything out item by item, not to mention having to reconstruct your lists one by one.

In [22]:
attack = data.loc["Bulbasaur": "Mew", "attack" ]
attack

name
Bulbasaur         49
Ivysaur           62
Venusaur          82
Mega Venusaur    100
Charmander        52
                ... 
Dragonite        134
Mewtwo           110
Mega Mewtwo X    190
Mega Mewtwo Y    150
Mew              100
Name: attack, Length: 197, dtype: int64

In [23]:
attack > 100

name
Bulbasaur        False
Ivysaur          False
Venusaur         False
Mega Venusaur    False
Charmander       False
                 ...  
Dragonite         True
Mewtwo            True
Mega Mewtwo X     True
Mega Mewtwo Y     True
Mew              False
Name: attack, Length: 197, dtype: bool

In [24]:
condition = attack > 100
attack[condition]

name
Mega Charizard X    130
Mega Charizard Y    104
Mega Blastoise      103
Mega Beedrill       150
Nidoking            102
Primeape            105
Arcanine            110
Machamp             130
Victreebel          105
Golem               120
Alolan Golem        120
Dodrio              110
Muk                 105
Alolan Muk          105
Krabby              105
Kingler             130
Alolan Exeggutor    105
Hitmonlee           120
Hitmonchan          105
Rhydon              130
Mega Kangaskhan     125
Scyther             110
Pinsir              125
Mega Pinsir         155
Gyarados            125
Mega Gyarados       155
Flareon             130
Kabutops            115
Aerodactyl          105
Mega Aerodactyl     135
Snorlax             110
Galarian Zapdos     125
Dragonite           134
Mewtwo              110
Mega Mewtwo X       190
Mega Mewtwo Y       150
Name: attack, dtype: int64

In [25]:
attack[attack > 100]

name
Mega Charizard X    130
Mega Charizard Y    104
Mega Blastoise      103
Mega Beedrill       150
Nidoking            102
Primeape            105
Arcanine            110
Machamp             130
Victreebel          105
Golem               120
Alolan Golem        120
Dodrio              110
Muk                 105
Alolan Muk          105
Krabby              105
Kingler             130
Alolan Exeggutor    105
Hitmonlee           120
Hitmonchan          105
Rhydon              130
Mega Kangaskhan     125
Scyther             110
Pinsir              125
Mega Pinsir         155
Gyarados            125
Mega Gyarados       155
Flareon             130
Kabutops            115
Aerodactyl          105
Mega Aerodactyl     135
Snorlax             110
Galarian Zapdos     125
Dragonite           134
Mewtwo              110
Mega Mewtwo X       190
Mega Mewtwo Y       150
Name: attack, dtype: int64

## More Filtering

Take a look at what's being done, and try to figure it out, particularly when it comes to the two-condition criteria!

In [27]:
# Can you figure out what's being done in the below code?

pokeAll = pd.read_csv("https://raw.githubusercontent.com/iamamangosteen/pythonforbusiness/master/pokedex_(Update_2021.04).csv", index_col = "name")
#import the data, set the index_col

criteria = (pokeAll["generation"] == 1) & (pokeAll["total_points"] > 100)
#define two filtering conditions (generation==1 and total_points>100) and save them in the criteria variable
#this is the first time we see column-shorthand notation

pokeAll["all_defense"] = pokeAll["defense"] + pokeAll["sp_defense"]
#create a new column called all_defense that is created by adding the defense and special defense columns

pokeAll[criteria][["defense","sp_defense", "all_defense"]]
#apply the filter criteria, then use column-shorthand notation to display only 3 columns

defense  sp_defense  all_defense
name                                           
Bulbasaur           49          65          114
Ivysaur             63          80          143
Venusaur            83         100          183
Mega Venusaur      123         120          243
Charmander          43          50           93
...                ...         ...          ...
Dragonite           95         100          195
Mewtwo              90          90          180
Mega Mewtwo X      100         100          200
Mega Mewtwo Y       70         120          190
Mew                100         100          200

[192 rows x 3 columns]

### <font color="red">Exercise 3: Pikachu data and filtering

* Find the speed of Raichu
* Find all the Pokemon in Generation 1 whose speed is greater than Raichu
    
There should be 17 Pokemon.

Bonus challenge: Filter out all "Mega" pokemon from the results.

In [28]:
raichu_speed = data.loc['Raichu']['speed']
print("Raichu's speed:", raichu_speed)
print()

condition = (data["generation"] == 1) & (data['speed'] > raichu_speed)
data[condition]['speed']

Raichu's speed: 110



name
Mega Beedrill      145
Mega Pidgeot       121
Partner Pikachu    120
Dugtrio            120
Persian            115
Alolan Persian     115
Alakazam           120
Mega Alakazam      150
Mega Gengar        130
Electrode          150
Starmie            115
Jolteon            130
Aerodactyl         130
Mega Aerodactyl    150
Mewtwo             130
Mega Mewtwo X      130
Mega Mewtwo Y      140
Name: speed, dtype: int64

In [29]:
#Bonus:
#can google for a solution and introduce stackoverflow as a good resource
condition = (data["generation"] == 1) & (data['speed'] > raichu_speed) & (~data.index.str.contains("Mega"))
data[condition]['speed']

name
Partner Pikachu    120
Dugtrio            120
Persian            115
Alolan Persian     115
Alakazam           120
Electrode          150
Starmie            115
Jolteon            130
Aerodactyl         130
Mewtwo             130
Name: speed, dtype: int64

### <font color="red">Exercise 4: What does this do?

What do each of the lines in this chunk of code do? Run it, find out, and explain to someone sitting next to you.

In [30]:
first = pd.read_csv('https://raw.githubusercontent.com/iamamangosteen/pythonforbusiness/master/pokedex_(Update_2021.04).csv', index_col = 'name')
#import the data, set the index_col

second = first[first['generation'] == 1]
#filter the dataframe to only display generation 1 pokemon

third = second.drop('Pikachu')
#drop Pikachu from the dataset (for whatever reason)

fourth = third.drop('generation', axis = 1)
#drop the generation column (presumably because it's no longer needed)

fourth.to_csv('result.csv')
fourth.to_excel('result.xlsx')
#export results to CSV and xlsx

## Inserting data

Inserting column data into your DataFrames is straightforward. Just add it in:

In [31]:
data.head()

pokedex_number german_name        japanese_name  generation  \
name                                                                         
Bulbasaur                   1     Bisasam  フシギダネ (Fushigidane)           1   
Ivysaur                     2   Bisaknosp   フシギソウ (Fushigisou)           1   
Venusaur                    3    Bisaflor  フシギバナ (Fushigibana)           1   
Mega Venusaur               3    Bisaflor  フシギバナ (Fushigibana)           1   
Charmander                  4    Glumanda      ヒトカゲ (Hitokage)           1   

               status         species  type_number type_1  type_2  height_m  \
name                                                                          
Bulbasaur      Normal    Seed Pokémon            2  Grass  Poison       0.7   
Ivysaur        Normal    Seed Pokémon            2  Grass  Poison       1.0   
Venusaur       Normal    Seed Pokémon            2  Grass  Poison       2.0   
Mega Venusaur  Normal    Seed Pokémon            2  Grass  Poison       2.4   
Charmander     Normal  Lizard Pokémon            1   Fire     NaN       0.6   

               ...  against_ground  against_flying against_psychic  \
name           ...                                                   
Bulbasaur      ...             1.0             2.0             2.0   
Ivysaur        ...             1.0             2.0             2.0   
Venusaur       ...             1.0             2.0             2.0   
Mega Venusaur  ...             1.0             2.0             2.0   
Charmander     ...             2.0             1.0             1.0   

              against_bug against_rock  against_ghost  against_dragon  \
name                                                                    
Bulbasaur             1.0          1.0            1.0             1.0   
Ivysaur               1.0          1.0            1.0             1.0   
Venusaur              1.0          1.0            1.0             1.0   
Mega Venusaur         1.0          1.0            1.0             1.0   
Charmander            0.5          2.0            1.0             1.0   

               against_dark  against_steel  against_fairy  
name                                                       
Bulbasaur               1.0            1.0            0.5  
Ivysaur                 1.0            1.0            0.5  
Venusaur                1.0            1.0            0.5  
Mega Venusaur           1.0            1.0            0.5  
Charmander              1.0            0.5            0.5  

[5 rows x 49 columns]

In [32]:
rows_to_show = ["Pichu", "Pikachu"]
cols_to_show = ['species', 'total_points', 'speed']
subset1 = data.loc[rows_to_show, cols_to_show]

rows_to_show = ["Raichu"]
cols_to_show = ['species', 'total_points', 'speed']
subset2 = data.loc[rows_to_show, cols_to_show]

rows_to_show = ["Raichu"]
cols_to_show = ['attack', 'defense']
subset3 = data.loc[rows_to_show, cols_to_show]

In [33]:
subset1

species  total_points  speed
name                                            
Pichu    Tiny Mouse Pokémon           205     60
Pikachu       Mouse Pokémon           320     90

In [34]:
subset2

species  total_points  speed
name                                      
Raichu  Mouse Pokémon           485    110

In [35]:
subset3

attack  defense
name                   
Raichu      90       55

In [36]:
subset1.append(subset2)

<ipython-input-36-ca7e2eaf0961>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  subset1.append(subset2)


species  total_points  speed
name                                            
Pichu    Tiny Mouse Pokémon           205     60
Pikachu       Mouse Pokémon           320     90
Raichu        Mouse Pokémon           485    110

In [37]:
subset1.append(subset3)

<ipython-input-37-cdb20573786e>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  subset1.append(subset3)


species  total_points  speed  attack  defense
name                                                             
Pichu    Tiny Mouse Pokémon         205.0   60.0     NaN      NaN
Pikachu       Mouse Pokémon         320.0   90.0     NaN      NaN
Raichu                  NaN           NaN    NaN    90.0     55.0

In [38]:
subset4 = data.loc["Bulbasaur": 'Charmander', "total_points": "defense"]
subset5 = data.loc["Bulbasaur": 'Charmander', "sp_attack": "speed"]

In [39]:
subset4.merge(subset5, left_index=True, right_index=True).head()

total_points  hp  attack  defense  sp_attack  sp_defense  speed
name                                                                          
Bulbasaur               318  45      49       49         65          65     45
Ivysaur                 405  60      62       63         80          80     60
Venusaur                525  80      82       83        100         100     80
Mega Venusaur           625  80     100      123        122         120     80
Charmander              309  39      52       43         60          50     65

### <font color="red">Exercise 5: Final Pokemon data analysis</font>

Write code to display each of the following:

* The Japanese names of all pokemon
* All data for the pokemon Charizard
* Total points for all pokemon in Generation 1.
* Total points for all Electric pokemon for all generations. (Hint: Type 1: Electric and Type 2: Electric both count)
* Attack points for all pokemon in Generation 4 as a multiple of special attack (sp_attack) for that pokemon. Show a DataFrame with 'attack', 'sp_attack', and "attack vs sp_attack", for example:

| name | attack | sp_attack | attack vs sp_attack |
|---|:---:|:---:|:---:|
| Turtwig | 68 | 45 | 1.51 |
| Grotle | 89 | 55 | 1.62 |
| Torterra | 109 | 75 | 1.45 |
| Chimchar | 58 | 58 | 1.00 |
| ... | ... | ... | ... |

In [40]:
data['japanese_name']

name
Bulbasaur               フシギダネ (Fushigidane)
Ivysaur                  フシギソウ (Fushigisou)
Venusaur                フシギバナ (Fushigibana)
Mega Venusaur           フシギバナ (Fushigibana)
Charmander                  ヒトカゲ (Hitokage)
                               ...         
Glastrier                ブリザポス (Burizaposu)
Spectrier                 レイスポス (Reisuposu)
Calyrex                     バドレックス (Budrex)
Calyrex Ice Rider           バドレックス (Budrex)
Calyrex Shadow Rider        バドレックス (Budrex)
Name: japanese_name, Length: 1045, dtype: object

In [41]:
data.loc["Charizard", :]

pokedex_number                     6
german_name                   Glurak
japanese_name       リザードン (Lizardon)
generation                         1
status                        Normal
species                Flame Pokémon
type_number                        2
type_1                          Fire
type_2                        Flying
height_m                         1.7
weight_kg                       90.5
abilities_number                   2
ability_1                      Blaze
ability_2                        NaN
ability_hidden           Solar Power
total_points                     534
hp                                78
attack                            84
defense                           78
sp_attack                        109
sp_defense                        85
speed                            100
catch_rate                      45.0
base_friendship                 70.0
base_experience                240.0
growth_rate              Medium Slow
egg_type_number                    2
e

In [42]:
condition = data['generation'] == 1
data[condition]['total_points']

name
Bulbasaur        318
Ivysaur          405
Venusaur         525
Mega Venusaur    625
Charmander       309
                ... 
Dragonite        600
Mewtwo           680
Mega Mewtwo X    780
Mega Mewtwo Y    780
Mew              600
Name: total_points, Length: 192, dtype: int64

In [44]:
criteria = (data['type_1'] == 'Electric') | (data['type_2'] == 'Electric')
data[criteria]['total_points']

name
Pikachu                    320
Partner Pikachu            430
Raichu                     485
Alolan Raichu              485
Alolan Geodude             300
                          ... 
Morpeko Full Belly Mode    436
Morpeko Hangry Mode        436
Dracozolt                  505
Arctozolt                  505
Regieleki                  580
Name: total_points, Length: 73, dtype: int64

In [45]:
subset1 = data[data['type_1'] == 'Electric']['total_points']
subset2 = data[data['type_2'] == 'Electric']['total_points']

subset1.append(subset2)

<ipython-input-45-c6c01dac5411>:4: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  subset1.append(subset2)


name
Pikachu            320
Partner Pikachu    430
Raichu             485
Alolan Raichu      485
Magnemite          325
                  ... 
Galvantula         472
Stunfisk           471
Zekrom             680
Charjabug          400
Vikavolt           500
Name: total_points, Length: 73, dtype: int64

In [46]:
data["attack vs sp_attack"] = round(data['attack'] / data['sp_attack'], 2)

condition = data['generation'] == 4
cols_to_show = ['attack', 'sp_attack', 'attack vs sp_attack']
data[condition][cols_to_show]

attack  sp_attack  attack vs sp_attack
name                                                      
Turtwig                 68         45                 1.51
Grotle                  89         55                 1.62
Torterra               109         75                 1.45
Chimchar                58         58                 1.00
Monferno                78         78                 1.00
...                    ...        ...                  ...
Manaphy                100        100                 1.00
Darkrai                 90        135                 0.67
Shaymin Land Forme     100        100                 1.00
Shaymin Sky Forme      103        120                 0.86
Arceus                 120        120                 1.00

[121 rows x 3 columns]